File: sdks/typescript/src/frame-builder.ts
/**
 * Frame Builder for ATP Protocol
 */

import { Frame, Payload, CompletionRequest } from './types';

export class FrameBuilder {
  private msgSeqCounters: Map<string, number> = new Map();

  constructor(
    private sessionId: string,
    private tenantId: string
  ) {}

  /**
   * Build a completion request frame
   */
  buildCompletionFrame(streamId: string, request: CompletionRequest, msgSeq?: number): Frame {
    if (msgSeq === undefined) {
      msgSeq = this.getNextMsgSeq(`${this.sessionId}:${streamId}`);
    }

    // Map quality to QoS
    const qosMap = {
      high: 'gold' as const,
      balanced: 'silver' as const,
      low: 'bronze' as const,
    };
    const qos = qosMap[request.quality || 'balanced'];

    // Build payload
    const payload: Payload = {
      type: 'completion',
      content: {
        prompt: request.prompt,
        maxTokens: request.maxTokens || 512,
        quality: request.quality || 'balanced',
        latencySloMs: request.latencySloMs || 5000,
        temperature: request.temperature || 0.7,
        stream: request.stream !== false,
        conversationId: request.conversationId,
        consistencyLevel: request.consistencyLevel || 'EVENTUAL',
      },
    };

    const frame: Frame = {
      v: 1,
      sessionId: this.sessionId,
      streamId,
      msgSeq,
      fragSeq: 0,
      flags: [],
      qos,
      ttl: 8,
      window: {
        maxParallel: 4,
        maxTokens: 50000,
        maxUsdMicros: 1000000,
      },
      meta: {
        taskType: 'completion',
        environmentId: this.tenantId,
      },
      payload,
    };

    return frame;
  }

  /**
   * Build a heartbeat frame
   */
  buildHeartbeatFrame(streamId: string): Frame {
    const msgSeq = this.getNextMsgSeq(`${this.sessionId}:${streamId}`);

    const payload: Payload = {
      type: 'heartbeat',
      content: {},
    };

    const frame: Frame = {
      v: 1,
      sessionId: this.sessionId,
      streamId,
      msgSeq,
      fragSeq: 0,
      flags: [],
      qos: 'bronze',
      ttl: 8,
      window: {
        maxParallel: 4,
        maxTokens: 50000,
        maxUsdMicros: 1000000,
      },
      meta: {
        taskType: 'heartbeat',
        environmentId: this.tenantId,
      },
      payload,
    };

    return frame;
  }

  /**
   * Get next message sequence number for a stream
   */
  private getNextMsgSeq(streamKey: string): number {
    const current = this.msgSeqCounters.get(streamKey) || 0;
    const next = current + 1;
    this.msgSeqCounters.set(streamKey, next);
    return next;
  }

  /**
   * Get current message sequence number for a stream
   */
  getCurrentMsgSeq(streamId: string): number {
    const streamKey = `${this.sessionId}:${streamId}`;
    return this.msgSeqCounters.get(streamKey) || 0;
  }

  /**
   * Reset message sequence counter for a stream
   */
  resetMsgSeq(streamId: string): void {
    const streamKey = `${this.sessionId}:${streamId}`;
    this.msgSeqCounters.delete(streamKey);
  }
}

File: go-sdk/frame_builder.go
package atpsdk

import (
	"encoding/json"
	"fmt"
	"time"
)

// FrameBuilder handles construction of ATP protocol frames
type FrameBuilder struct {
	sessionID string
	tenantID  string
	msgSeqCounters map[string]int
}

// NewFrameBuilder creates a new frame builder
func NewFrameBuilder(sessionID, tenantID string) *FrameBuilder {
	return &FrameBuilder{
		sessionID:      sessionID,
		tenantID:       tenantID,
		msgSeqCounters: make(map[string]int),
	}
}

// getNextMsgSeq returns the next message sequence number for a stream
func (fb *FrameBuilder) getNextMsgSeq(streamID string) int {
	key := fmt.Sprintf("%s:%s", fb.sessionID, streamID)
	fb.msgSeqCounters[key]++
	return fb.msgSeqCounters[key]
}

// BuildCompletionFrame builds a completion request frame
func (fb *FrameBuilder) BuildCompletionFrame(streamID string, request CompletionRequest) Frame {
	msgSeq := fb.getNextMsgSeq(streamID)

	return Frame{
		Type:      "completion_request",
		Timestamp: time.Now().UnixMilli(),
		StreamID:  streamID,
		MsgSeq:    msgSeq,
		FragSeq:   0,
		Flags:     []string{},
		QoS:       "gold",
		TTL:       8,
		Window: Window{
			MaxParallel: 4,
			MaxTokens:   50000,
			MaxUSD:      1000000,
		},
		Meta: Meta{
			TaskType:      "completion",
			EnvironmentID: fb.tenantID,
		},
		Payload: map[string]interface{}{
			"prompt":      request.Prompt,
			"max_tokens":  request.MaxTokens,
			"temperature": request.Temperature,
			"top_p":       request.TopP,
			"stop":        request.Stop,
		},
	}
}

// BuildHeartbeatFrame builds a heartbeat frame
func (fb *FrameBuilder) BuildHeartbeatFrame() Frame {
	return Frame{
		Type:      "heartbeat",
		Timestamp: time.Now().UnixMilli(),
		Payload:   map[string]interface{}{},
	}
}

// BuildCapabilityFrame builds a capability advertisement frame
func (fb *FrameBuilder) BuildCapabilityFrame(streamID string, capability CapabilityAdvertisement) Frame {
	msgSeq := fb.getNextMsgSeq(streamID)

	return Frame{
		Type:      "adapter.capability",
		Timestamp: time.Now().UnixMilli(),
		StreamID:  streamID,
		MsgSeq:    msgSeq,
		FragSeq:   0,
		Flags:     []string{"capability"},
		QoS:       "bronze",
		TTL:       30, // Longer TTL for capability frames
		Window: Window{
			MaxParallel: 1,
			MaxTokens:   1000,
			MaxUSD:      10000,
		},
		Meta: Meta{
			EnvironmentID: fb.tenantID,
		},
		Payload: map[string]interface{}{
			"type":                 "adapter.capability",
			"adapter_id":           capability.AdapterID,
			"adapter_type":         capability.AdapterType,
			"capabilities":         capability.Capabilities,
			"models":               capability.Models,
			"max_tokens":           capability.MaxTokens,
			"supported_languages":  capability.SupportedLanguages,
			"cost_per_token_micros": capability.CostPerTokenMicros,
			"health_endpoint":      capability.HealthEndpoint,
			"version":              capability.Version,
			"metadata":             capability.Metadata,
		},
	}
}

// BuildHealthFrame builds a health status frame
func (fb *FrameBuilder) BuildHealthFrame(streamID string, health HealthStatus) Frame {
	msgSeq := fb.getNextMsgSeq(streamID)

	return Frame{
		Type:      "adapter.health",
		Timestamp: time.Now().UnixMilli(),
		StreamID:  streamID,
		MsgSeq:    msgSeq,
		FragSeq:   0,
		Flags:     []string{"health"},
		QoS:       "bronze",
		TTL:       60, // Health frames have longer TTL
		Window: Window{
			MaxParallel: 1,
			MaxTokens:   1000,
			MaxUSD:      10000,
		},
		Meta: Meta{},
		Payload: map[string]interface{}{
			"type":                 "adapter.health",
			"adapter_id":           health.AdapterID,
			"status":               health.Status,
			"p95_latency_ms":       health.P95LatencyMS,
			"p50_latency_ms":       health.P50LatencyMS,
			"p99_latency_ms":       health.P99LatencyMS,
			"requests_per_second":  health.RequestsPerSecond,
			"error_rate":           health.ErrorRate,
			"queue_depth":          health.QueueDepth,
			"memory_usage_mb":      health.MemoryUsageMB,
			"cpu_usage_percent":    health.CPUUsagePercent,
			"uptime_seconds":       health.UptimeSeconds,
			"version":              health.Version,
			"last_health_check":    time.Now().Unix(),
			"metadata":             health.Metadata,
		},
	}
}

// SerializeFrame serializes a frame to JSON bytes
func (fb *FrameBuilder) SerializeFrame(frame Frame) ([]byte, error) {
	return json.Marshal(frame)
}

// DeserializeFrame deserializes JSON bytes to a frame
func (fb *FrameBuilder) DeserializeFrame(data []byte) (Frame, error) {
	var frame Frame
	err := json.Unmarshal(data, &frame)
	return frame, err
}
